File: 백준/Silver/1497. 기타콘서트/기타콘서트.py
import sys

input = sys.stdin.readline


def solve(curr, currFlag, count):
    if curr == n:
        c = str(bin(currFlag)[2:]).count("1")
        return (c, count)

    results = []
    results.append(solve(curr + 1, currFlag, count))

    _, flag = guitars[curr]
    for i in range(m):
        if flag[i] == "Y":
            currFlag |= 1 << i

    results.append(solve(curr + 1, currFlag, count + 1))

    results.sort(key=lambda x: (-x[0], x[1]))
    return results[0]


n, m = map(int, input().split())
guitars = [list(input().split()) for _ in range(n)]

answer = solve(0, 0, 0)

if answer[0] == 0:
    print(-1)
else:
    print(answer[1])

File: 백준/Silver/1497. 기타콘서트/기타콘서트.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int n, m, maxSong, minGuitar;
  static String[][] guitars;

  public static void main(String[] args) throws IOException {
    setUp();

    maxSong = 0;
    minGuitar = Integer.MAX_VALUE;

    subset(0, 0, new boolean[m]);

    if (maxSong == 0) {
      sb.append(-1);
    } else {
      sb.append(minGuitar);
    }

    output();
  }

  private static void subset(int curr, int count, boolean[] flag) {
    if (curr == n) {
      int songCount = 0;
      for (int i = 0; i < m; i++) {
        if (flag[i]) {
          songCount++;
        }
      }

      if (maxSong < songCount) {
        maxSong = songCount;
        minGuitar = count;
      } else if (maxSong == songCount) {
        minGuitar = Math.min(minGuitar, count);
      }
      return;
    }

    subset(curr + 1, count, flag);

    boolean[] currFlag = new boolean[m];
    for (int i = 0; i < m; i++) {
      if (flag[i] || guitars[curr][1].charAt(i) == 'Y') {
        currFlag[i] = true;
      }
    }
    subset(curr + 1, count + 1, currFlag);

  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());
    guitars = new String[n][2];
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      guitars[i][0] = st.nextToken();
      guitars[i][1] = st.nextToken();
    }
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }

}